modules/json/json-parser.c

```diff
@@ -265,6 +292,11 @@ json_parser_extract(JSONParser *self, struct json_object *jso, LogMessage *msg)
       json_parser_process_object(jso, self->prefix, msg);
       return TRUE;
     }
+  if (json_object_is_type(jso, json_type_array))
+    {
+      json_parser_process_array(jso, self->prefix, msg);
```

Comment:
Adding top-level JSON array support in a way that we parse them into `$1, $2, $3` sounds good, but this implementation suggests top-level arrays are not really supported out-of-the-box, as matches are special nv-pairs that, for example, are not transmitted over network when using a basic $(format-json) template. `$1, $2, $3` sounds like we want to support this only in SCL blocks where these can be processed further.

I'm just thinking, but wouldn't it be better to do the same with the top-level arrays as we do with embedded arrays (creating a $(list))? We could use `prefix()` as the name of the list.
If `prefix` is not set, we could still fall back to `$1, $2...` or just use a special hard-coded name for the list.

Reply:
I agree that just putting these into matches is not much, however mapping this list to $prefix does not really help much either.

My solution was to implement `$*` as a macro in a followup patch that produces a list from the numbered matches, which makes it much easier to process this further. I don't think the array form is to be considered a final representation of the message, thus it shouldn't be added as a "real" name-value pair.

This is the patch that implements `$*`: https://github.com/syslog-ng/syslog-ng/pull/3831/commits/2131b60f60bef4294be3539bde6685286071320b

We could discuss this IRL perhaps.

Reply:
Thank you, it would be great to discuss this. 
We had a few thoughts on this topic at the weekly OSE meeting.

Reply:
We've discussed this IRL on the weekly OSE meeting, it was a long and complicated topic with lots of pros and cons.
We agreed to go with the "$1, $2, ..." method with top-level arrays for now because there were multiple issues with the out-of-the box support "nameless", "top-level" entities in syslog-ng (format-json, value-pairs, etc.).

This special case does not break anything in v3.xx, only adds experimental support for top-level arrays. We can change the details of it until 4.0.